fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/synonym/SynonymAnalyzer.java

```diff
@@ -33,30 +35,61 @@
 import org.apache.lucene.analysis.core.FlattenGraphFilter;
 import org.apache.lucene.analysis.standard.StandardAnalyzer;
 import org.apache.lucene.analysis.standard.StandardTokenizer;
+import org.apache.lucene.analysis.synonym.SolrSynonymParser;
 import org.apache.lucene.analysis.synonym.SynonymGraphFilter;
 import org.apache.lucene.analysis.synonym.SynonymMap;
-import org.apache.lucene.analysis.synonym.WordnetSynonymParser;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
 import java.text.ParseException;
+import java.util.Objects;
 
 /**
  * The analyzer for index with synonym enabled.
  * Only for in fly analysis during query time.
  * The synonyms are not indexed to disk. So no index rebuilding is needed if the wordnet file gets updated.
  */
 public class SynonymAnalyzer extends StopwordAnalyzerBase {
```

Comment:
Are `SynonymAnalyzer`s thread safe (at least for reading from)? I ask just because if they're going to be cached in the registry, then they might be used by multiple threads.

I also wonder if this might be worth separating into an abstract `SynonymAnalyzer` class with abstract methods for getting the stop words and/or synonym stream, and then putting the "default" stuff (like the file names and stop word list) in a `DefaultSynonymAnalyzer` class. That way, it's clear to other implementors what they need to do to produce their own analyzers.

Reply:
@alecgrieser if I understand this PR correctly, only the ```synonymMap``` could potentially be shared by different threads and it is OK because it is ```final```, and threads using the same analyzer are supposed to use same map. @arnaud-lacurie please correct me if I am wrong.

Reply:
@alecgrieser, I looked more into it and I think you are right. `SynonymMaps` are thread safe, but there is no guarantee that `Analyzers` are. I introduced a `SynonymMap` registry instead.

Reply:
Cool, cool, sounds good

> if I understand this PR correctly, only the synonymMap could potentially be shared by different threads and it is OK because it is `final`

For the record, I think `final` isn't enough here, because a `final` map can still be mutated and, importantly here, its elements can still have mutable internal state. The only thing `final` guarantees is that you can't change the map pointer to point to something else, which is better than nothing, but not as much as maybe we'd like.